framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceJavaParserStorage.java

```diff
@@ -936,30 +936,82 @@ private String addClassesForElement(Element element) {
    *
+   * <p>Because of the side effect, clients may want to pass a copy into this method.
+   *
    * @param classAnnos the class annotations to modify
+   * @param supertypes the binary names of all supertypes; not side-effected
+   * @param subtypes the binary names of all subtypes; not side-effected
    */
-  public void wpiPrepareClassForWriting(ClassOrInterfaceAnnos classAnnos) {
+  public void wpiPrepareClassForWriting(
+      ClassOrInterfaceAnnos classAnnos,
+      Collection<@BinaryName String> supertypes,
+      Collection<@BinaryName String> subtypes) {
+    if (classAnnos.callableDeclarations.isEmpty()) {
+      return;
+    }
+
     for (Map.Entry<String, CallableDeclarationAnnos> methodEntry :
         classAnnos.callableDeclarations.entrySet()) {
-      wpiPrepareMethodForWriting(methodEntry.getValue());
+      String jvmSignature = methodEntry.getKey();
+      List<CallableDeclarationAnnos> inSupertypes =
+          findOverloads(jvmSignature, supertypesMap.get(classAnnos.className));
+      List<CallableDeclarationAnnos> inSubtypes =
+          findOverloads(jvmSignature, subtypesMap.get(classAnnos.className));
+
+      wpiPrepareMethodForWriting(methodEntry.getValue(), inSupertypes, inSubtypes);
+    }
+  }
+
+  /**
+   * Return all the CallableDeclarationAnnos for the given signature.
+   *
+   * @param jvmSignature the JVM signature
+   * @param typeNames a collection of type names
+   * @return the CallableDeclarationAnnos for the given signature, in all of the types
+   */
+  private List<CallableDeclarationAnnos> findOverloads(
+      String jvmSignature, Collection<@BinaryName String> typeNames) {
+    if (typeNames == null) {
+      return Collections.emptyList();
+    }
+    List<CallableDeclarationAnnos> result = new ArrayList<>();
+    for (String typeName : typeNames) {
+      ClassOrInterfaceAnnos classAnnos = classToAnnos.get(typeName);
+      if (classAnnos != null) {
+        CallableDeclarationAnnos callableAnnos = classAnnos.callableDeclarations.get(jvmSignature);
+        if (callableAnnos != null) {
+          result.add(callableAnnos);
+        }
+      }
     }
+    return result;
   }
 
   /**
    * Side-effects the method or constructor annotations to make any desired changes before writing
    * to a file.
    *
    * @param methodAnnos the method or constructor annotations to modify
+   * @param inSupertypes the method or constructor annotations for all overridden methods; not
+   *     side-effected
+   * @param inSubtypes the method or constructor annotations for all overriding methods; not
+   *     side-effected
    */
-  public void wpiPrepareMethodForWriting(CallableDeclarationAnnos methodAnnos) {
-    atypeFactory.wpiPrepareMethodForWriting(methodAnnos);
+  // TODO: The type factory implementation must consider both inferred annotations *and* those
+  // written by the programmer.  Handle annotations in elements later.
```

Comment:
Can you expand on this comment a bit? I think the point it is making is the following, but it's too terse for me to be sure:
* the current implementation of `wpiPrepareMethodForWriting` only considers annotations that were inferred in this round of WPI (because it is based on `CallableDeclarationAnnos`s)
* in theory, a type factory might want to make changes to the inferred annotations on a method based on either the inferred annotations or the programmer written annotations (i.e., "annotations in elements")
* currently, this isn't supported, because there's no way to get from a `CallableDeclarationAnnos` to an element (are we sure about this? `CallableDeclaration`, at least, stores the `ExecutableElement` for the method, so an enterprising type system developer may already be able to accomplish this in a roundabout way)
* it would be desirable to support considering also the annotations written by the programmer (accessed via the relevant elements), but you're leaving it for later

Is that right?